add ConTopSpikeColors

ByFox213 opened the pull request:
Closed: #307

ChillerDragon commented on src/game/server/instagib/chat_commands.cpp:
```diff
@@ -483,6 +483,56 @@ void CGameContext::ConRankFlagCaptures(IConsole::IResult *pResult, void *pUserDa
+	if(!pSelf->m_pController)
+		return;
+
+	if(pSelf->m_pController->IsDDRaceGameType())
```
This should check for fng not for ddrace gametype. There are no spikes in zCatch or ctf for example.

ChillerDragon commented on src/game/server/instagib/chat_commands.h:
```diff
@@ -58,5 +58,6 @@ CONSOLE_COMMAND("rank_flags", "?r[player name]", CFGFLAG_CHAT | CFGFLAG_SERVER,
 CONSOLE_COMMAND("top5flags", "?i[rank to start with]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConTopFastcaps, this, "Shows the all time best ranks by flag time")
 CONSOLE_COMMAND("top5caps", "?i[rank to start with]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConTopNumCaps, this, "Shows the all time best ranks by amount of flag captures")
 CONSOLE_COMMAND("rank_caps", "?r[player name]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConRankFlagCaptures, this, "Shows the all time flag capture rank of player name (your stats by default)")
+CONSOLE_COMMAND("top5spikes", "?r[color] ?i[rank to start with]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConTopSpikeColors, this, "Shows the best ranks in terms of spike kills of all time")
```
``r`` means rest or remaining. It will consume even spaces. It should be ``s`` for space separated string.
Otherwise ``/top5spikes green 5`` will receive the color ``green 5`` instead of ``green``. Also the possible values can already be listed there like this

```
CONSOLE_COMMAND("top5spikes", "?s['gold'|'green'|'purple'] ?i[rank to start with]",
```

Also maybe without the ``?`` for the color? Not sure but looking at the code the ux seems a bit tricky. If a user types in /top5spikes into the chat they just get a list of all the colors without any instruction on what to do or where to put it. If it was not optional it would show the usage when the color is missing. Otherwise if you want a custom usage message by making it optional it should be more obvious to the user that they have to provide a color as argument.

ChillerDragon replied:
The wording ``"Shows the best ranks in terms of spike kills of all time"`` is a bit different than the other top commands but I guess its okay.

ByFox213 replied:
> The wording `"Shows the best ranks in terms of spike kills of all time"` is a bit different than the other top commands but I guess its okay.

I'll change it to this message
```
Shows the all time best ranks by spike kills
```

ChillerDragon commented on src/game/server/instagib/chat_commands.cpp:
```diff
@@ -483,6 +483,56 @@ void CGameContext::ConRankFlagCaptures(IConsole::IResult *pResult, void *pUserDa
+
+	for(const char *pColor : aSpikeColors)
+	{
+		if(strcmp(SpikeColor, pColor) == 0)
```
should be ``str_comp_nocase`` instead of ``strcmp``

ChillerDragon commented on src/game/server/instagib/chat_commands.cpp:
```diff
@@ -483,6 +483,56 @@ void CGameContext::ConRankFlagCaptures(IConsole::IResult *pResult, void *pUserDa
+	const char *pName = pSelf->Server()->ClientName(pResult->m_ClientId);
+	const char *SpikeColor = pResult->GetString(0);
+	const int Offset = pResult->NumArguments() > 1 ? pResult->GetInteger(1) : 1;
+	const char *aSpikeColors[] = {
```
Should be ``apSpikeColors`` not ``aSpikeColors`` it is an array of pointers.

ByFox213 replied:
Forgot to rename it

ChillerDragon commented on src/game/server/instagib/chat_commands.cpp:
```diff
@@ -483,6 +483,56 @@ void CGameContext::ConRankFlagCaptures(IConsole::IResult *pResult, void *pUserDa
+	}
+
+	const char *pName = pSelf->Server()->ClientName(pResult->m_ClientId);
+	const char *SpikeColor = pResult->GetString(0);
```
Missing p prefix for pointer ``pSpikeColor``

ChillerDragon commented on src/game/server/instagib/chat_commands.cpp:
```diff
@@ -483,6 +483,56 @@ void CGameContext::ConRankFlagCaptures(IConsole::IResult *pResult, void *pUserDa
+	pSelf->SendChatTarget(pResult->m_ClientId, "~~~ Spike colors");
+	for(const char *pColor : aSpikeColors)
+	{
+		pSelf->SendChatTarget(pResult->m_ClientId, ("~ " + std::string(pColor)).c_str());
```
Please don't use ``std::string`` use ``str_format`` instead to stay in the tw code style.

ChillerDragon reviewed (changes requested):
Thanks for working on this

ByFox213 committed:
feat: Add spike kill rankings with /top5spikes

Implement all-time kill statistics for gold/green/purple spikes with ranking filters.

Closes #307

ByFox213 committed:
feat: Add spike kill rankings with /top5spikes

Implement all-time kill statistics for gold/green/purple spikes with ranking filters.

Closes #307

ByFox213 commented:
```
if(pSelf->m_pController->IsFngGameType())
```
to
```
if(!pSelf->m_pController->IsFngGameType())
```

ChillerDragon commented:
Thanks
